### PR TITLE
Migrate Steps to Go modules if GOPATH is unsupported

### PR DESCRIPTION
**Pull request by lpusok**
### Checklist

- Requires  MINOR  [version update](https://semver.org/)

### Context

Support Go 1.16's default setting and later Go releases that no longer support GOPATH and have the go env GO111MODULE set to "on". We migrate older Step releases to Go modules on the fly, during executing the workflow.

Resolves: https://bitrise.atlassian.net/browse/STEP-670

### CODE REVIEW

File: toolkits/golang.go [hunk 1]
```diff
@@ -255,48 +255,67 @@ func (toolkit GoToolkit) Install() error {
...
+		log.Debugf("[Go deps] Step requires GOPATH mode")
+		// Go 1.17 will ignore GO111MODULE (https://blog.golang.org/go116-module-changes)
+		// GO111MODULE needs to be set to "on" when GOPATH is no longer supported.
+		// If GO111MODULE is not set, will be handled as it was "on".
+		mode, err := getGoEnv(cmdRunner, goConfig.GoBinaryPath, "GO111MODULE")
+		if err != nil {
+			log.Warnf("[Go deps] Could not determine if GOPATH mode is supported: %v", err)
+		}
+
+		if isGoPathModeSupported(mode) {
```
Comment (istvankovacs-bitrise) [inline toolkits/golang.go:271]: Are you sure that the `GO111MODULE` will be removed in Go 1.17 and not just ignored?

Comment (lpusok) [reply toolkits/golang.go:271]: According to https://blog.golang.org/go116-module-changes it will be ignored. We will have to set it to `on`, which is handled the same as the env was removed.

File: toolkits/golang.go [hunk 2]
```diff
@@ -337,16 +356,23 @@ func (toolkit GoToolkit) PrepareForStepRun(step stepmanModels.StepModel, sIDData
...
 	if step.Toolkit == nil {
 		return errors.New("No Toolkit information specified in step")
 	}
 	if step.Toolkit.Go == nil {
 		return errors.New("No Toolkit.Go information specified in step")
 	}
-	packageName := step.Toolkit.Go.PackageName
 
-	return goBuildInIsolation(packageName, stepAbsDirPath, fullStepBinPath)
+	isInstallRequired, _, goConfig, err := selectGoConfiguration()
```
Comment (istvankovacs-bitrise) [inline toolkits/golang.go:366]: Could you please elaborate on the reason why did you move it here?

File: toolkits/modmigrate.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	return err != nil
+}
+
+func migrateToGoModules(stepAbsDirPath, packageName string) error {
+	if packageName == "" {
+		return fmt.Errorf("package name not specified")
+	}
+
+	goModTemplate := `module %s
+go 1.16`
```
Comment (istvankovacs-bitrise) [inline toolkits/modmigrate.go:56]: Why did you hardcode the version here? Did you consider fetching the current Go version instead?

File: toolkits/golang.go (hunk 2 above)
Comment (lpusok) [reply toolkits/golang.go:366]: Wanted to test goBuildStep and the go path on the actual machine would have affected the command run, this was a cheap fix for that.

File: toolkits/golang.go (hunk 1 above)
Comment (istvankovacs-bitrise) [reply toolkits/golang.go:271]: Discussed offline, the env will be just ignored, but we have to manually set to `on` on the VMs.

File: toolkits/go_cmd_builder.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+func (g goCmdBuilder) goBuildEnv(stepAbsDir string, shouldCheckGoSum bool) []string {
+	// GOSUMDB=off disables requests to checksum databases (https://golang.org/ref/mod#communicating-with-proxies)
+	// There is no go.sum file as migrated go modules are in use,
+	// so it is used only for an extra measure disabling network queries.
+	envs := []string{
+		"PWD=" + stepAbsDir,
+		"GOROOT=" + g.goConfig.GOROOT,
+		"GO111MODULE=on",
+	}
+	if shouldCheckGoSum {
```
Comment (istvankovacs-bitrise) [inline toolkits/go_cmd_builder.go:32]: An `!` is missing.

File: toolkits/go_cmd_builder.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+		envs = append(envs, "GOSUMDB=off")
+	}
+
+	return envs
+}
+
+func (g goCmdBuilder) goBuildMigratedModules(stepAbsDir string, outputBin string) *command.Model {
+	buildCmd := command.New(g.goConfig.GoBinaryPath, g.goBuildArgs(outputBin, true)...).
+		AppendEnvs(g.goBuildEnv(stepAbsDir, false)...).
```
Comment (istvankovacs-bitrise) [inline toolkits/go_cmd_builder.go:41]: This should be `true`.

File: toolkits/go_cmd_builder.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+		SetDir(stepAbsDir)
+
+	return buildCmd
+}
+
+func (g goCmdBuilder) goBuildInModuleMode(stepAbsDir string, outputBin string) *command.Model {
+	buildCmd := command.New(g.goConfig.GoBinaryPath, g.goBuildArgs(outputBin, false)...).
+		AppendEnvs(g.goBuildEnv(stepAbsDir, true)...).
```
Comment (istvankovacs-bitrise) [inline toolkits/go_cmd_builder.go:49]: This should be `false`.

File: toolkits/go_cmd_builder.go (hunk 4 above)
Comment (lpusok) [reply toolkits/go_cmd_builder.go]: Fixed.

**Review by istvankovacs-bitrise — APPROVED** [review]
